### DISCUSSION TITLE
[Fixed][@lit-labs/router] Uncaught (in promise) TypeError: this.t is not iterable at i$1.goto (add a playground)

### DISCUSSION THREAD

**Discussion by ynxtynxt** [lit/lit#2935] [General]
~~Update: I can't reproduce this anymore since @lit-labs/router 0.1.0~~
Update: [Fixed](https://github.com/lit/lit/issues/3031#issuecomment-1158982271)

Add a playground [ ** here ** ](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy5qcyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7IFJvdXRlciB9IGZyb20gXCJodHRwczovL2Nkbi5za3lwYWNrLmRldi9AbGl0LWxhYnMvcm91dGVyXCI7XG5cbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcblxuICBzdGF0aWMgcHJvcGVydGllcyA9IHtcbiAgICBuYW1lOiB7dHlwZTogU3RyaW5nfSxcbiAgfTtcblxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMucm91dGVyID0gbmV3IFJvdXRlcih0aGlzLCBbe1xuICAgICAgcGF0aDogJyonLCAvLyB0cmlnZ2VyIHRhaWxHcm91cFxuICAgICAgcmVuZGVyOiAoKSA9PiBodG1sYCR7d2luZG93LmxvY2F0aW9uLnBhdGhuYW1lfTxicj5cbiAgICAgICAgUGxlYXNlIGNoZWNrIGNvbnNvbGU8YnI-XG4gICAgICAgIDxidXR0b24gQGNsaWNrPSR7IGUgPT4gdGhpcy5yb3V0ZXIuZ290bygnLycpfT5SZXByb2R1Y2U8L2J1dHRvbj5cbiAgICAgICAgYCxcbiAgICB9XSlcbiAgfVxuXG4gIHJlbmRlcigpIHtcbiAgICAgcmV0dXJuIGh0bWxgJHt0aGlzLnJvdXRlci5vdXRsZXQoKX1gO1xuICB9XG59XG5jdXN0b21FbGVtZW50cy5kZWZpbmUoJ3NpbXBsZS1ncmVldGluZycsIFNpbXBsZUdyZWV0aW5nKTtcbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPjwvc2ltcGxlLWdyZWV0aW5nPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

It's still there, I guess I didn't test properly. To reproduce ```router.goto()``` should be called and use npm package and not /development.

A quick observation (I'm not really a web-dev so I don't really understand JS that well),
```node_modules/@lit-labs/router/routes.js``` has ```this.t=[]```
```node_modules/@lit-labs/router/router.js``` has ```this.t=t=>{...```
It looks like ```this.t``` was overrided by subclass? Or I didn't load the module properly? Should I file a bug report if it isn't my fault?

I could do with some help, please.. In my app I have...

```js
  _router = new Router(this, [
    {path: '/', render: () => html`<h1>Home</h1>`},
    {path: '/dish/*', render: () => html`<x-child></x-child>`},
  ]);
```

And I have

```js
@customElement('x-child')
class XChild extends LitElement {
  _routes = new Routes(this, [
    {path: 'cuisine', render: () => html`<h1>cuisine</h1>`},
  ]);
  render() {
    return html`${this._routes.outlet()}`;
  }
}
```

Then I entered url `http://localhost:5000/dish/cuisine` and got `Uncaught (in promise) TypeError: this.t is not iterable`.
```js
        if (void 0 !== n)
            for (const t of this.t)
                t.goto(n);
        this.l.requestUpdate();
```
from the implementation of `async goto()` so I put a `console.log(this.t)` before the `for(...` it seems to be a function

```
t=>{const i=0!==t.button||t.metaKey||t.ctrlKey||t.shiftKey;if(t.defaultPrevented||i)return;const s=t.composedPath().find(t=>"A"===t.tagName);if(void 0===s||""!==s.target||s.hasAttribute("download")||…
```

Which looks like it came from

```js
export class Router extends Routes {
    constructor() {
        super(...arguments);
        this._onClick = (e) => {
            const isNonNavigationClick = e.button !== 0 || e.metaKey || e.ctrlKey || e.shiftKey;
            if (e.defaultPrevented || isNonNavigationClick) {
                return;
            }

```

I won't have problem if I 
```js
import { Routes, Router } from '@lit-labs/router/development';
```
instead of 
```js
import { Routes, Router } from '@lit-labs/router';
```

some how it seems to me

```ts
export class Routes implements ReactiveController {
//...
 private readonly _childRoutes: Array<Routes> = [];

// ie. this.t=[] in routes.js
```

got overwritten by 
```ts
export class Router extends Routes {
//...
private _onClick = (e: MouseEvent) => {
    const isNonNavigationClick =
      e.button !== 0 || e.metaKey || e.ctrlKey || e.shiftKey;
    if (e.defaultPrevented || isNonNavigationClick) {
      return;
    }
// ie. this.t=t=>{const i=0!==t.button||t.metaKey in router.js
```
So if I change all `this.t` to `this.tt` it seems to work just fine.

Could someone give me an idea why this is happening? Thank you a lot in advance.

**Comment by drk-mtr** [comment]
This appears to work:

```ts
  router = new Router(this, [
    { path: '/', render: () => html`Other component` },
    {
      path: '/__playground_swfs_f56081d9/2uk68gr/index.html',
      render: () => html`${window.location.pathname}<br>
        Please check console<br>
        <button @click=${ e => this.router.goto('/')}>Reproduce</button>
        `,
    }
  ])
```

From the README.md:

> Routes can be nested: a route path can include a trailing /* pattern to match against a prefix, and will automatically propagate that prefix to Routes controllers defined in child elements.

**Reply by ynxtynxt** [reply]
Thanks! Though you will need to have a root * (for serving 404) or a nested child route like /children/* to trigger the `tailGroup`. Else it will work just fine. The playground was the min to reproduce the case.

https://github.com/lit/lit/blob/7ba356c4795ddf2b2a69e69a4c3f78df1e6b56c5/packages/labs/router/src/routes.ts#L202-L206

`_childRoutes` will be overrided by `_onClick` after minified as they both converted into `this.t`
